### PR TITLE
[clang][frontend] Add support for attribute plugins for statement attributes

### PR DESCRIPTION
**Pull request by ericastor**
We already have support for declaration attributes; this is just a matter of extending the plugin infrastructure to cover one more case.

### CODE REVIEW

**Comment by ilya-biryukov** [comment]
This looks like a relatively straightforward extension in line with the current Clang interfaces. Therefore, I feel we should accept this.

However, please wait for feedback from Clang owners to make sure this does not go against some non-obvious trade-offs or future plans that Clang has.

cc @AaronBallman, @cor3ntin

**Review by Sirraide — COMMENTED** [review]
Looks ok to me, but I’ll leave it to @erichkeane to approve this as he’s the attributes code owner.

**Review by erichkeane — COMMENTED** [review]
So the problem with statement attributes is that we don't really have a good way to do instantiation of them on a template, which is why we held off on this in the first place. The infrastructure for instantiation DOES happen on decl attributes automatically anyway, so that made it easier to do.

So please ensure there is some sort of hooks in the statement attributes to ensure that they get a chance to instantatiate themselves (and add tests with templates in them!).

**Comment by ericastor** [comment]
> So the problem with statement attributes is that we don't really have a good way to do instantiation of them on a template, which is why we held off on this in the first place. The infrastructure for instantiation DOES happen on decl attributes automatically anyway, so that made it easier to do.
> 
> So please ensure there is some sort of hooks in the statement attributes to ensure that they get a chance to instantatiate themselves (and add tests with templates in them!).

This is a fair point... how are standard statement attributes handled inside of templates? (Assuming clang supports that.)

**Comment by erichkeane** [comment]
> > So the problem with statement attributes is that we don't really have a good way to do instantiation of them on a template, which is why we held off on this in the first place. The infrastructure for instantiation DOES happen on decl attributes automatically anyway, so that made it easier to do.
> > So please ensure there is some sort of hooks in the statement attributes to ensure that they get a chance to instantatiate themselves (and add tests with templates in them!).
> 
> This is a fair point... how are standard statement attributes handled inside of templates? (Assuming clang supports that.)

All our handling of attributes is pretty manual for anything with arguments.  We have a bit of an automatic handling for many attributes, but arguments aren't.  I think there is a section in `SemaInstantiateDecl.cpp` that handles a lot of it (`InstantiateAttrs`) and you can see there is a LOT of manual handling there

**Comment by ericastor** [comment]
So... I think this turned out to be surprisingly easy. After all, we don't actually allow custom attributes to create entirely new attributes in the AST, since they can't create new legal AST entries... instead, the example creates AnnotateAttrs that can then be observed & acted on by the compiler-extending plugin.

It looks like reaching the same level of support just required adding AnnotateAttr support to the template instantiation logic. Am I missing anything else? @erichkeane

File: clang/lib/Sema/SemaTemplateInstantiate.cpp
Comment (erichkeane) [inline clang/lib/Sema/SemaTemplateInstantiate.cpp:2138]: we probably want to transform the annotation as well?  or does `StringArgument` not work for dependent values?

**Review by erichkeane — COMMENTED** [review]
This seems reasonable?  I actually didn't know we weren't letting them create new attributes, and were just using `AnnotateAttr` in the AST.  I think @AaronBallman should do 1 run through this as he understands the plugins better than I, but this otherwise LGTM.

File: clang/lib/Sema/SemaTemplateInstantiate.cpp
Comment (ericastor) [reply clang/lib/Sema/SemaTemplateInstantiate.cpp:2138]: I think `StringArgument` only handles `StringRef`s? I don't see a way it could be given a dependent value.

**Review by AaronBallman — APPROVED** [review]
LGTM, though we should probably add a release note to clang/docs/ReleaseNotes.rst so users know that the plugin system got a bit more functionality.

**Pull request merged**
